### PR TITLE
Add optional event debouncing for auto-restart

### LINKED ISSUE #542: watchdog runs too many times / repeatedly when many files are changed
I am using watchdog to watch a git repository and rerun a build process every time the repo changes.

I have observed that performing a git pull, for example, can cause watchdog to run repeatedly.  I *believe* that the git pull is causing a very large number of files to change, and hence watchdog runs many times because watchdog runs once per file modification.

For me, this is not the desired behavior.  If 5 files are modified at the same time, for example, I only need watchdog to run the build process once.  I am hoping we can add a configuration option to watchdog to allow somebody to opt-in to the following proposed behavior "lazy run":

## Proposed behavior "lazy run":

  * Watchdog maintains a queue of file change events.  This happens concurrently.

  1. Check the queue.
  2. If the queue is nonempty, empty the queue and run the process.
  3. (When the process is finished,) return to step 1.

## Example of "lazy run" in action:

Suppose that the process to run takes a long time, and 5 files are modified in rapid succession.

  1. First file modified.
  2. modification 1 goes in queue
  3. watchdog empties queue and begins running the process
  4. Second file modified
  5. Third file modified
  6. Fourth file modified
  7. Fifth file modified
  8. modifications 2 through 5 are in the queue.
  9. The process completes and watchdog checks the queue again
  10. Modifications are discarded from the queue and watchdog begins running the process again
  11. The process completes again.

As you can see, in this example, even though 5 files were modified, the process only runs 2 times.

### PR DESCRIPTION
**Pull request by taleinat**
If a debounce interval is specified, upon receiving events, collect events until the interval passes with no incoming events, and only then trigger a restart.

Closes #542.

Note to reviewers: The more involved start and stop mechanisms in AutoRestartTrick are needed to avoid restarting the debouncer when restart the process, and to ensure that the debouncer is properly stopped when stopping the trick.

### CODE REVIEW

File: tests/test_0_watchmedo.py [hunk 1]
```diff
@@ -106,6 +106,43 @@ def test_shell_command_subprocess_termination_nowait(tmpdir):
...
+    trick.on_any_event("foo/bar.baz")
+    trick.on_any_event("foo/bar2.baz")
+    time.sleep(0.1)
+    trick.on_any_event("foo/bar3.baz")
+    time.sleep(1)
+    trick.on_any_event("foo/bar.baz")
+    time.sleep(1)
+    trick.stop()
+    cap = capfd.readouterr()
+    assert cap.out.splitlines(keepends=False).count('+++++ 0') == 3
```
Comment (BoboTiG) [inline tests/test_0_watchmedo.py:143]: Maybe add a comment about what changes are expected?

File: src/watchdog/tricks/__init__.py [hunk 2]
```diff
@@ -177,60 +179,106 @@ class AutoRestartTrick(Trick):
 
     def __init__(self, command, patterns=None, ignore_patterns=None,
                  ignore_directories=False, stop_signal=signal.SIGINT,
-                 kill_after=10):
+                 kill_after=10, debounce_interval_seconds=0):
         super().__init__(
             patterns=patterns, ignore_patterns=ignore_patterns,
             ignore_directories=ignore_directories)
+
+        if kill_after < 0:
```
Comment (BoboTiG) [inline src/watchdog/tricks/__init__.py:187]: Those checks could be moved above the previous line.

**Comment by BoboTiG** [comment]
Thanks a lot @taleinat!

Would you mind adding a line in the changelog too, please?

**Comment by taleinat** [comment]
> Would you mind adding a line in the changelog too, please?

Of course, done.

File: tests/test_0_watchmedo.py (hunk 1 above)
Comment (taleinat) [reply tests/test_0_watchmedo.py:152]: I've added doc-strings to the auto-restart tests.

File: src/watchdog/tricks/__init__.py (hunk 2 above)
Comment (taleinat) [reply src/watchdog/tricks/__init__.py:183]: Done.

File: tests/test_0_watchmedo.py [hunk 3]
```diff
@@ -106,17 +106,68 @@ def test_shell_command_subprocess_termination_nowait(tmpdir):
...
```
Comment (BoboTiG) [inline tests/test_0_watchmedo.py:152]:
```suggestion
    assert cap.out.splitlines(keepends=False).count('+++++ 0') == 3
    assert trick.restart_count == 2
```

File: src/watchdog/tricks/__init__.py [hunk 4]
```diff
@@ -177,60 +179,108 @@ class AutoRestartTrick(Trick):
...
-        self.process = subprocess.Popen(self.command, preexec_fn=getattr(os, 'setsid', None))
-        self.process_watcher = ProcessWatcher(self.process, self._restart)
-        self.process_watcher.start()
+        self.event_debouncer = EventDebouncer(self.debounce_interval_seconds, lambda events: self._restart_process())
+        self.event_debouncer.start()
+        self._start_process()
 
     def stop(self):
-        if self.process is None:
+        # Ensure the body of the function is only run once.
```
Comment (oprypin) [inline src/watchdog/tricks/__init__.py:212]: Why do we need to ensure that it runs only once? It seems like it should behave OK even if called multiple times.

File: src/watchdog/utils/event_debouncer.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+            if self.debounce_interval_seconds:
+                # Collect additional events until the debounce interval passes.
+                while self.should_keep_running():
+                    try:
+                        events.append(self._event_queue.get(timeout=self.debounce_interval_seconds))
+                    except queue.Empty:
+                        break
+
+            self.events_callback(events)
```
Comment (oprypin) [inline src/watchdog/utils/event_debouncer.py:47]: Here's a different implementation, like I did it [in MkDocs](https://github.com/mkdocs/mkdocs/blob/56b235a8ad43f2300d17f87e6fa4de7a3d764397/mkdocs/livereload/__init__.py#L166).

It has the benefit of being instantly stoppable, rather than having a 0 - 0.2 sec delay before being able to stop.

It also ensures to instantly stop firing events when it's stopped, maybe  eliminating the need for
> The more involved start and stop mechanisms in AutoRestartTrick are needed to avoid restarting the debouncer when restart the process

```suggestion
import logging
import threading


logger = logging.getLogger(__name__)


class EventDebouncer(threading.Thread):
    """Background thread for debouncing event handling.

    When an event is received, wait until the configured debounce interval
    passes before calling the callback.  If additional events are received
    before the interval passes, reset the timer and keep waiting.  When the
    debouncing interval passes, the callback will be called with a list of
    events in the order in which they were received.
    """
    def __init__(self, debounce_interval_seconds, events_callback):
        super().__init__()
        self.debounce_interval_seconds = debounce_interval_seconds
        self.events_callback = events_callback

        self._cond = threading.Condition()
        self._shutdown = False
        self._events = []

    def handle_event(self, event):
        with self._cond:
            self._events.append(event)
            self._cond.notify_all()

    def run(self):
        while True:
            with self._cond:
                while not self._cond.wait_for(lambda: self._events or self._shutdown, timeout=0.2):
                    # We could have used just one wait instead of a loop + timeout, but we need
                    # occasional breaks, otherwise on Windows we can't receive KeyboardInterrupt.
                    pass
                if self._shutdown:
                    break
                if self.debounce_interval_seconds:
                    while self._cond.wait(timeout=self.debounce_interval_seconds):
                        # Wait for events to stop happening.
                        pass

                events = list(self._events)
                self._events.clear()

            self.events_callback(events)

    def stop(self):
        with self._cond:
            self._shutdown = True
            self._cond.notify_all()
```

**Comment by BoboTiG** [comment]
I'm wondering if we can get the debouncer for watchdog too 🤔 
Did you already think about such thing @taleinat?

**Comment by taleinat** [comment]
> I'm wondering if we can get the debouncer for watchdog too :thinking:

@BoboTiG could you explain what you mean by that? This is already implemented as a reusable `EventDebouncer` class.

Do you mean whether this could be made easier to use with Trick sub-classes, e.g. with AutoRestartTrick?

File: src/watchdog/utils/event_debouncer.py (hunk 5 above)
Comment (taleinat) [reply src/watchdog/utils/event_debouncer.py:47]: Thanks for this @oprypin!

Using a condition like this is indeed better for instant stopping, I'll work that into the implementation.

Comment (oprypin) [reply src/watchdog/utils/event_debouncer.py:47]: @taleinat Well you can just "accept suggestion", I think that will also end up simpler than, for example, keeping it a queue. It's a fully functioning replacement, just in 1 of the new tests will need to add a microscopic sleep to make it pass.

**Comment by oprypin** [comment]
@taleinat I'm guessing what was meant (or at least I think it would be good) if usages of `Observer` / `observer.schedule` could simply add a debounce parameter without needing to use a separate class. But you're right, your class can indeed satisfy this requirement. And maybe it's already the most idiomatic usage, but in that case an example of how to use it with an observer and your own callback would really make your idea shine.

**Comment by BoboTiG** [comment]
@taleinat sorry for thr lack of information :)
The idea from @oprypin is what I meant: introduce an easy way for users to use the new class. The keyword argument seems like a good idea.

File: src/watchdog/utils/event_debouncer.py (hunk 5 above)
Comment (taleinat) [reply src/watchdog/utils/event_debouncer.py:47]: Thanks, and your work to make it such a drop-in replacement was helpful and is appreciated!

This version however did not handle well the case where a shutdown signal is given during the debouncing interval. Also, the use of `Condition.wait_for()` is more complex than necessary.

So instead I've integrated the approach while addressing these two concerns.

**Comment by taleinat** [comment]
> @taleinat sorry for thr lack of information :) The idea from @oprypin is what I meant: introduce an easy way for users to use the new class. The keyword argument seems like a good idea.

I spent several hours looking into this, but I haven't found a reasonable way to achieve this with the current design of the watchdog codebase.

EventEmitter classes can emit many different types of events. Support the complex number of possible combination of which events to debounce, together or separately, seems like it would make things harder and more complex than using the EventDebouncer class directly. The same goes for EventDispatcher.

I think auto-restart is a bit of a special use-case, as it treats all events *from all files* identically. When implementing this, I did consider whether it would be good to also implement debouncing for other commands (especially shell-command), but I couldn't think of good use cases for that.

So, to summarize, IMO this change is immediately useful and the implementation is reasonable, so it should go in. Future improvements may be done in the future ;)

**Comment by BoboTiG** [comment]
You mark a point. Let's merge that great PR when I'll find some time on a real computer :)
